### PR TITLE
Make slide scrolling smoother by adding a continued scrolling velocity

### LINKED ISSUE #1058: Make slide scrolling smoother
### Describe the feature

We need to add support for acceleration and deceleration to slide scrolling, in addition to making it so that it continues to scroll after you stop sliding.

### Relevant code

_No response_

### PR DESCRIPTION
**Pull request by kkoreilly**
If we have enough velocity, we continue scrolling over the next second in a goroutine while slowly decelerating for a smoother experience.

Fixes #1058

### CODE REVIEW

**Comment by kkoreilly** [comment]
@rcoreilly can you test this on Android (web and/or native)? Also, do you think we need to call AsyncLock and AsyncUnlock around the scrolling? It seems to work both ways, and I don't think there should be any issues, but it might make sense to just be safe? Might that also have a negative performance impact though?

File: core/frame.go [hunk 1]
```diff
@@ -144,12 +144,34 @@ func (fr *Frame) Init() {
...
+		// smoother experience.
+		vel := math32.Vector2FromPoint(e.StartDelta()).DivScalar(float32(e.SinceStart().Milliseconds())).Negate()
+		if math32.Abs(vel.X) < 1 && math32.Abs(vel.Y) < 1 {
+			return
+		}
+		go func() {
+			i := 0
+			tick := time.NewTicker(time.Second / 60)
+			for range tick.C {
+				fr.scrollDelta(events.NewScroll(e.WindowPos(), vel, e.Modifiers()))
```
Comment (rcoreilly) [inline core/frame.go:166]: absolutely need AsyncLock / Unlock here.  time cost is minimal.

File: core/frame.go [hunk 2]
```diff
@@ -144,12 +144,34 @@ func (fr *Frame) Init() {
...
+				vel.SetMulScalar(0.9)
```
Comment (rcoreilly) [inline core/frame.go]: 0.9 is pretty rapid decay.. probably should at least have a gobal unexported for this.

File: events/source.go [hunk 3]
```diff
@@ -94,6 +94,12 @@ func (es *Source) KeyChord(rn rune, code key.Codes, mods key.Modifiers) {
 // MouseButton creates and sends a mouse button event with given values
 func (es *Source) MouseButton(typ Types, but Buttons, where image.Point, mods key.Modifiers) {
 	ev := NewMouse(typ, but, where, mods)
+	if es.Last.MouseButtonType == MouseDown {
```
Comment (rcoreilly) [inline events/source.go]: shouldn't this be conditional on Up?  you definitely don't want to just have all down events be the same!?

Comment (kkoreilly) [reply events/source.go]: What do you mean? How does this make them all the same?

File: core/frame.go (hunk 1 above)
Comment (kkoreilly) [reply core/frame.go:166]: Done.

File: core/frame.go (hunk 2 above)
Comment (kkoreilly) [reply core/frame.go]: No point in having a global unexported if you don't need it. Don't add premature complexity! We can add a setting later if we want to. I will experiment with other values.

File: events/source.go (hunk 3 above)
Comment (rcoreilly) [reply events/source.go]: You are copying from last always.

Comment (kkoreilly) [reply events/source.go]: I am just copying the start and prev info, which would otherwise be unset.

Comment (rcoreilly) [reply events/source.go]: Down should not have that info - only up.

Comment (kkoreilly) [reply events/source.go]: Done.

**Comment by rcoreilly** [comment]
Last one resolved prematurely on my phone... Just exclude down for current when setting this info from last.

**Comment by kkoreilly** [comment]
I already did.

File: core/frame.go (hunk 2 above)
Comment (kkoreilly) [reply core/frame.go]: What do you think about the speed now?

File: events/source.go [hunk 4]
```diff
@@ -94,6 +94,12 @@ func (es *Source) KeyChord(rn rune, code key.Codes, mods key.Modifiers) {
 // MouseButton creates and sends a mouse button event with given values
 func (es *Source) MouseButton(typ Types, but Buttons, where image.Point, mods key.Modifiers) {
 	ev := NewMouse(typ, but, where, mods)
+	if typ == MouseUp && es.Last.MouseButtonType == MouseDown {
```
Comment (rcoreilly) [inline events/source.go]: sorry to be a PITA but the key thing is to just *exclude* MouseDown -- MouseDrag definitely needs these things.

Comment (rcoreilly) [reply events/source.go]: or maybe that goes in a different place?  anyway, just exclude MouseDown to be clearer.

Comment (kkoreilly) [reply events/source.go]: You know what the great thing about MouseDrag is? It goes through MouseMove, not MouseButton. The only things that go through MouseButton are MouseDown and MouseUp, but if you really want me to invert the logic for you, I can.

Comment (kkoreilly) [reply events/source.go]: Good?

**Comment by kkoreilly** [comment]
The main remaining issue is that you can't get it to velocity scroll with a very small swipe due to the velocity threshold, but if you lower the threshold, then non-swipes incorrectly result in velocity scrolling.

**Comment by rcoreilly** [comment]
do you want to play with that more or should we merge this?

**Comment by kkoreilly** [comment]
Can you test it on Android and give feedback on speed and threshold?

**Comment by kkoreilly** [comment]
We can figure out the speed and threshold later; this is already a major improvement. @rcoreilly please merge.

**Review by rcoreilly — APPROVED** [review]

**Pull request merged**